keps/sig-storage/2605-option-to-preserve-default-file-mode-bit-in-atomicwriter-volumes/README.md

```diff
@@ -0,0 +1,170 @@
+    // relative and may not contain the '..' path or start with '..'.
+    // +optional
+    Items []KeyToPath
+    // Mode bits to use on created files by default. Must be a value between
+    // 0 and 0777.
+    // Directories within the path are not affected by this setting.
+    // This might be in conflict with other options that affect the file
+    // mode, like fsGroup, and the result can be other mode bits set.
+    // If PreserveDefaultMode is set to true then file mode bit is not
+    // modified after the file creation, due to any other options set
+    // such as fsGroup etc
+    // +optional
+    DefaultMode *int32
+    // Specify whether the Secret or its key must be defined
+    // +optional
+    Optional *bool
+    // Specify whether we want to preserve the DefaultMode set on the files
+    // at the time of file creation.
+    // If PreserveDefaultMode is set to true the file mode is not modified based
+    // on fsGroup after the file is created with the defaultMode.
+    // +optional
+    PreserveDefaultMode *bool
+}
+```
+
+A secret volume source with preserveDefaultMode=true:
+
+```yaml
+  volumes:
+  - name: test-secret
+    secret:
+      preserveDefaultMode: true
+      defaultMode: 256
+      secretName: test-secret
+```
+
+### File Permission
+
+For the volumes that use AtomicWriter today the files are created with the Mode set in the KeyToPath struct.
+If the Mode is not specified then the DefaultMode set for the source is used.
+If DefaultMode is not specified then the file is created with mode bit set to 0600.
+If pod have fsGroup specified then all the files in the volume are walked,
+chown and chmod to the fsGroup
+
+#### Proposed heuristics
+
+- *Case 1*: The volume has PreserveDefaultMode set to false.
+    In this case there is no behavioural change in how the file permissions are
+    set for the volume. In other words, we will continue to create the files
+    with Mode\Default and then set the volume ownership based on fsGroup if specified.
+    This is the default case.
+- *Case 2*: The volume has PreserveDefaultMode set to true.
+    In this case the payload\volume\files are created using Mode\Default mode.
+    Then the call to SetVolumeOwnership is skipped, in turn preserving the Default file mode on the files.
+
+#### Alternatives considered
+
+- Instead of having PreserveDefaultMode flag at the volume source level, we can have the same flag in
+    the KeyToPath struct. That will allow us granular control over every file with-in the volume :
+  - We would have to pass the payload information along with PreserveDefaultMode to the SetVolumeOwnership routine
+        and skip chown\chmod if PreserveDefaultMode is set to true for the file.
+  - This will be more invasive change, but would allow more granular control at per file level
+  - If users want to preserve Mode for one file in the volume and not for another they can achieve that by creating
```

Comment:
I think another alternative would have been to use field introduced in https://github.com/kubernetes/enhancements/issues/695 to respect `FSGroupChangePolicy` for secrets/configmaps/emptydirs too.

Reply:
FSGroupChangePolicy is a pod level setting whereas in our case we do want to only specify the atomicwriter volume file permission.

Reply:
Yeah, I think the difference here is that if user wants the behavior for a specific volume and not all volumes for the pod, which appears to be a common grievance in the discussion.

Reply:
Could there be a case where user is using secret volume with more than one container in pod but still wants to preserve mode bits? It would require either both containers to run with same UID or containers running as root.

My main concern is - do we need *per-volume* granularity? Because when we discussed this with @msau42 @tallclair  - while @tallclair proposed, it might be helpful to have *per-volume* granularity , in practical terms either pods that use fsgroup will want to apply it to all volumes or not. Also per-volume granularity is confusing for end users.

Reply:
@gnufied I think these are the ones which were being highlighted in the discussions. Take a look at this one. In certain scenarios users apparently want to preserve the fileMode bit set and that's the motivation here.
https://github.com/istio/istio/issues/32217

I am not sure if there are workarounds available for issues such as above. Just following the conversation and making an educated attempt to address.